hail/src/main/scala/is/hail/backend/service/Worker.scala

```diff
@@ -44,21 +69,114 @@ class WorkerTimer() {
+    if (x == null) "" else x
+  }
+  private[this] val workerGCSFS = retryTransientErrors {
+    using(new FileInputStream(s"/worker-key.json")) { is =>
+      new GoogleStorageFS(IOUtils.toString(is, Charset.defaultCharset().toString()))
+    }
+  }
+
+  private[this] def withThreadContextClassLoader[T](custom: ClassLoader)(f: => T): T = {
+    val original = Thread.currentThread().getContextClassLoader()
+    try {
+      Thread.currentThread().setContextClassLoader(custom)
+      f
+    } finally {
+      Thread.currentThread().setContextClassLoader(original)
+    }
+  }
+
+  private[this] val STRING_ARRAY_CLASS = Class.forName("[Ljava.lang.String;")
+  private[this] def loadMainFromLocalJar(revision: String, localJarURL: URL): Array[String] => Unit = {
+    val classLoader = new LoadSelfFirstURLClassLoader(Array(localJarURL))
+    val reflectedMain = withThreadContextClassLoader(classLoader) {
+      val workerClass = classLoader.loadClass("is.hail.backend.service.Worker")
+
+      val actualRevision = workerClass.getMethod("myRevision").invoke(null).asInstanceOf[String]
+      assert(revision == actualRevision, s"$revision != $actualRevision")
+
+      workerClass.getMethod("main", STRING_ARRAY_CLASS)
+    }
+
+    { (args: Array[String]) =>
+      withThreadContextClassLoader(classLoader) {
+        try {
+          reflectedMain.invoke(null, args): Unit
+        } catch {
+          case e: InvocationTargetException =>
+            throw e.getCause()
+        }
+      }
+    }
+  }
+
+  private[this] def downloadFile(remoteFS: FS, remote: String, local: String): Unit =
+    retryTransientErrors {
+      using(remoteFS.openNoCompression(remote)) { is =>
+        using(new FileOutputStream(local)) { os =>
+          IOUtils.copy(is, os)
+        }
+      }
+    }
+
+  private[this] def mainForRevision(revision: String, remoteJarLocation: String): Array[String] => Unit = {
+    revisions.get(revision) match {
+      case Some(mainMethod) =>
+        log.info(s"$revision found in the memory cache")
+        mainMethod
+      case None =>
+        log.info(s"$revision not in the memory cache")
+
+        val localJarLocation = s"/hail-jars/${revision}.jar"
+        val localJarFile = new File(localJarLocation)
+        if (!localJarFile.exists()) {
+          log.info(s"$revision not in the disk cache")
+          localJarFile.getParentFile().mkdirs()
+          downloadFile(workerGCSFS, remoteJarLocation, localJarLocation)
+          log.info(s"$revision added to the disk cache at $localJarLocation")
+        }
+
+        val localJarURL = localJarFile.toURI().toURL()
+        val mainMethod = loadMainFromLocalJar(revision, localJarURL)
+        log.info(s"$revision loaded")
+        revisions.put(revision, mainMethod)
+        mainMethod
     }
-    val root = args(0)
-    val i = args(1).toInt
+  }
+
+  def main(args: Array[String]): Unit = {
     val timer = new WorkerTimer()
+    timer.start("main")
```

Comment:
Will this start two `main` timings when an alternate jar is loaded? I think this span is a bit confusing

Reply:
Yeah this is a bit confusing. Suppose we have the Worker who is first called `A` and the worker we reflect into `B`. `A`'s timer is basically abandoned. I never call end, so I never print out any information about that timing. `B`'s main method creates a new timer that knows nothing about the other timer (in fact, it probably can't refer to anything in `A` since it believes it is the one, true is.hail.backend.service.Worker).

I suppose a fix would be to break main like this:

```
  def main(args: Array[String]): Unit = {
    val timer = new WorkerTimer()
    timer.start("main")

    log.info(s"is.hail.backend.service.Worker $myRevision")
    if (args.length < 2) {
      throw new IllegalArgumentException(s"expected at least two arguments, not: ${ args.length }")
    }

    val revision = args(0)
    val remoteJarLocation = args(1)
    if (revision != myRevision) {
      timer.start("JAR fetching and Worker class loading")
      log.info(s"received job for different revision: $revision; I am $myRevision")
      val runMethod = runForRevision(revision, remoteJarLocation)
      timer.end("JAR fetching and Worker class loading")
      runMethod(args)
      timer.end("main")
    } else {
      run(args)
      timer.end("main")
    }
  }

  def run(args: Array[String]): Unit = {
    /* only Hail execution stuff, no class loading */
  }
```

Reply:
I think this is _a lot_ easier to follow